minichain/chain.py

```
@@ -35,29 +29,29 @@ def last_block(self):
-                logger.warning("Block %s rejected: Invalid previous hash %s != %s", block.index, block.previous_hash, self.last_block.hash)
+                logger.warning("Block %s rejected: Invalid previous hash", block.index)
                 return False
 
             # Check index linkage
             if block.index != self.last_block.index + 1:
-                logger.warning("Block %s rejected: Invalid index %s != %s", block.index, block.index, self.last_block.index + 1)
+                logger.warning("Block %s rejected: Invalid index", block.index)
                 return False
 
             # Verify block hash
             if block.hash != calculate_hash(block.to_header_dict()):
```

Ruff (0.15.2)

[error] Undefined name `calculate_hash`

(F821)